src/protocols/dhcpv6/decode.c

```diff
@@ -438,26 +439,34 @@ static ssize_t decode_option(TALLOC_CTX *ctx, fr_cursor_t *cursor, fr_dict_t con
 
 	if ((da->type == FR_TYPE_STRING) && da->flags.subtype) {
 		rcode = decode_dns_labels(ctx, cursor, dict, da, data + 4, len, decoder_ctx);
-
 	} else if (da->flags.array) {
 		rcode = decode_array(ctx, cursor, dict, da, data + 4, len, decoder_ctx);
-
 	} else if (da->type == FR_TYPE_VSA) {
 		rcode = decode_vsa(ctx, cursor, dict, da, data + 4, len, decoder_ctx);
-
 	} else if (da->type == FR_TYPE_TLV) {
 		rcode = decode_tlvs(ctx, cursor, dict, da, data + 4, len, decoder_ctx);
-
 	} else {
 		rcode = decode_value(ctx, cursor, dict, da, data + 4, len, decoder_ctx);
 	}
 
 	if (rcode < 0) return rcode;
 
+	/*
+	 *	It is a hack!
+	 *
+	 *	RFC 4704 says "FQDN", unless it's a single label, in which case it's a
+	 *	partial name, and we omit the trailing zero.
+	 */
+	for (vp = fr_cursor_head(cursor); vp; vp = fr_cursor_next(cursor)) {
+		if (vp->da->type == FR_TYPE_STRING &&
+			vp->da->flags.subtype == FLAG_ENCODE_PARTIAL_DNS_LABEL) {
+			fr_pair_value_bstrncpy(vp, vp->vp_strvalue+1, vp->vp_length-1);
+		}
+	}
+
```

Comment:
This decoding really belongs in the `fr_struct_from_network()` function.  It should have a `decode_value()` callback, just like the `encode_value()` callback in `fr_struct_to_network()`.

That way it works for arrays of DNS labels inside of a struct, too